protocols/autonatv2/src/client/behaviour.rs

```diff
@@ -81,148 +51,240 @@ where
+                });
+            }
         }
-        todo!()
+        if let Some(dial_request) = self.pending_requests.pop_front() {
+            if self.known_servers.is_empty() {
+                self.pending_requests.push_front(dial_request);
+            } else {
+                let peer = self.known_servers.choose(&mut self.rng).unwrap();
+
+                self.submit_req_for_peer(*peer, dial_request);
+            }
+        }
+        Poll::Pending
     }
 }
 
 impl<R> Behaviour<R>
 where
-    R: RngCore,
+    R: RngCore + 'static,
 {
-    fn handle_response_info(
-        &mut self,
-        peer_id: PeerId,
-        ResponseInfo {
-            response:
-                DialResponse {
-                    status,
-                    dial_status,
-                    ..
-                },
-            suspicious_addrs,
-            successfull_addr,
-        }: ResponseInfo,
-    ) -> Result<ReachabilityTestSucc, ReachabilityTestErr> {
-        match (status, dial_status) {
-            (ResponseStatus::E_REQUEST_REJECTED, _) => {
-                Err(ReachabilityTestErr::ServerRejectedDialRequest { peer_id })
-            }
-            (ResponseStatus::E_DIAL_REFUSED, _) => {
-                Err(ReachabilityTestErr::ServerChoseNotToDialAnyAddress { peer_id })
-            }
-            (ResponseStatus::E_INTERNAL_ERROR, _) => {
-                Err(ReachabilityTestErr::InternalServerError { peer_id })
-            }
-            (ResponseStatus::OK, DialStatus::UNUSED) => {
-                Err(ReachabilityTestErr::InvalidResponse { peer_id })
-            }
-            (ResponseStatus::OK, DialStatus::E_DIAL_ERROR) => {
-                Err(ReachabilityTestErr::UnableToConnectOnSelectedAddress {
-                    peer_id,
-                    addr: successfull_addr,
-                })
-            }
-            (ResponseStatus::OK, DialStatus::E_DIAL_BACK_ERROR) => {
-                Err(ReachabilityTestErr::FailureDuringDialBack {
-                    peer_id,
-                    addr: successfull_addr,
-                })
-            }
-            (ResponseStatus::OK, DialStatus::OK) => Ok(ReachabilityTestSucc {
-                server_peer: peer_id,
-                visible_addr: successfull_addr,
-                suspicious_addrs,
-            }),
+    fn submit_req_for_peer(&mut self, peer: PeerId, req: DialRequest) {
+        if let Some(conn_id) = self.peers_to_handlers.get(&peer) {
+            self.pending_events.push_back(ToSwarm::NotifyHandler {
+                peer_id: peer,
+                handler: NotifyHandler::One(*conn_id),
+                event: Either::Left(RequestFromBehaviour::PerformRequest(req)),
+            });
+        } else {
+            self.pending_events.push_back(ToSwarm::Dial {
+                opts: DialOpts::peer_id(peer)
+                    .condition(PeerCondition::DisconnectedAndNotDialing)
+                    .build(),
+            });
```

Comment:
I think this is unnecessary. I would suggest that the autonat protocol never makes connections by itself. Instead, we should just piggy-back off connections made by other protocols or the user.

This means we should only send autonat requests to server that a) we confirmed that they support the protocol b) we are still connected to.

Reply:
That means if we got disconnected from a server that supports the protocol, we should just remove that server from our list?

Reply:
I think so. This also means we avoid somewhat unbounded growth of this list. Imagine a node that is running for a long time and crawls the network and keeps discovering autonat servers.

Reply:
I think that is resolved.